File: 0130-surrounded-regions/0130-surrounded-regions.cs
public class Solution {
    public void Solve(char[][] board) {
        // intuition: start traversal (bfs or dfs) from 0s that are on the boundary and find all connected zeros four directionally. Preserve these zeros as they are not surrounded by x on atleast one side. Also maintain a visited array to ensure that no zero is visited twice.
        // using bfs
        int n=board.Length;
        int m = board[0].Length;
        int[,] visited = new int[n,m];
        if(n==1 && m==1)
        {
            return;
        }
        Queue<Tuple<int,int>> bfs = new();
        for(int i=0; i<n; i++)
        {
            if(board[i][0] == 'O')
            {
                bfs.Enqueue(Tuple.Create(i,0));
                visited[i,0]=1;
            }
            if(board[i][m-1] == 'O')
            {
                bfs.Enqueue(Tuple.Create(i,m-1));
                visited[i,m-1]=1;
            }
        }
        for(int j=1; j<m-1; j++)
        {
            if(board[0][j] == 'O')
            {
                bfs.Enqueue(Tuple.Create(0,j));
                visited[0,j]=1;
            }
            if(board[n-1][j] == 'O')
            {
                bfs.Enqueue(Tuple.Create(n-1,j));
                visited[n-1,j]=1;
            }
        }
        int[] deltaRow = [1, 0, -1, 0];
        int[] deltaCol = [0, 1, 0, -1];
        while(bfs.Count > 0)
        {
            var top = bfs.Dequeue();
            int row = top.Item1;
            int col = top.Item2;
            for(int k=0; k<4; k++)
            {
                int newRow = row + deltaRow[k];
                int newCol = col + deltaCol[k];
                if(newRow < n && newRow >= 0 && newCol < m && newCol >= 0 && visited[newRow, newCol] == 0 && board[newRow][newCol] == 'O')
                {
                    bfs.Enqueue(Tuple.Create(newRow,newCol));
                    visited[newRow,newCol] = 1;
                }
            }
        }
        for(int i=0; i<n; i++)
        {
            for(int j=0; j<m; j++)
            {
                if(visited[i,j] == 1)
                {
                    board[i][j] = 'O';
                }
                else
                {
                    board[i][j] = 'X';
                }
            }
        }
    }
}

File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    public void solve(char[][] board) {
        int r = board.length;
        int c = board[0].length;
        Queue<int[]> bfsQueue = new LinkedList<>();
        boolean[][] vis = new boolean[r][c];
        // top
        for(int i=0; i<c; i++)
        {
            if(board[0][i] == 'O' && !vis[0][i])
            {
                bfsQueue.add(new int[] { 0, i });
                vis[0][i] = true;
            }
        }

        // bottom
        for(int i=0; i<c; i++)
        {
            if(board[r-1][i] == 'O' && !vis[r-1][i])
            {
                bfsQueue.add(new int[] { r-1, i });
                vis[r-1][i] = true;
            }
        }

        // left
        for(int i=0; i<r; i++)
        {
            if(board[i][0] == 'O' && !vis[i][0])
            {
                bfsQueue.add(new int[] { i, 0 });
                vis[i][0] = true;
            }
        }

        // left
        for(int i=0; i<r; i++)
        {
            if(board[i][c-1] == 'O' && !vis[i][c-1])
            {
                bfsQueue.add(new int[] { i, c-1 });
                vis[i][c-1] = true;
            }
        }
        int[][] DIRS = {{0,1}, {0,-1}, {-1,0}, {1,0}};
        while(bfsQueue.size() > 0)
        {
            var top = bfsQueue.poll();
            int i = top[0];
            int j = top[1];
            for(int[] d : DIRS)
            {
                int nr = i + d[0];
                int nc = j + d[1];
                if(nr < 0 || nr == r || nc < 0 || nc == c || vis[nr][nc] || board[nr][nc] == 'X')
                {
                    continue;
                }
                bfsQueue.add(new int[] { nr, nc });
                vis[nr][nc] = true;
            }
        }
        for(int i=0; i<r; i++)
        {
            for(int j=0; j<c; j++)
            {
                if(!vis[i][j])
                {
                    board[i][j] = 'X';
                }
            }
        }
    }
}